### PR TITLE
fix: avoid out of bounds error when rendering short hashes

### PR DESCRIPTION
**Pull request by Stebalien**
The panic is caught higher up the stack so it can't crash the node, but it's still "not good".

### CODE REVIEW

**Comment by aschmahmann** [comment]
I think the problem is slightly different. It looks like `hash=""` https://github.com/ipfs/go-ipfs/blob/cab67f6b6635b53bbf32680ee0da982e32bdcfee/core/corehttp/gateway_handler.go#L394 and when there's an error we just continue. What's going on there?

**Comment by Stebalien** [comment]
Ah, I see. Yeah, that makes no sense.

**Comment by Stebalien** [comment]
I fixed it both ways.

File: core/corehttp/gateway_handler.go [hunk 1]
```diff
@@ -391,11 +391,12 @@ func (i *gatewayHandler) getOrHeadHandler(w http.ResponseWriter, r *http.Request
...
-		hash := ""
-		if r, err := i.api.ResolvePath(r.Context(), ipath.Join(resolvedPath, dirit.Name())); err == nil {
-			// Path may not be resolved. Continue anyways.
-			hash = r.Cid().String()
+		resolved, err := i.api.ResolvePath(r.Context(), ipath.Join(resolvedPath, dirit.Name()))
+		if err != nil {
+			internalWebError(w, err)
+			return
 		}
+		hash := resolved.Cid().String()
```
Comment (aschmahmann) [inline core/corehttp/gateway_handler.go:399]: @lidel what was the idea behind "Path may not be resolved. Continue anyways.", it seems like we should just fail here as this PR does?

**Review by aschmahmann — APPROVED** [review]
Thanks for the quick fix, it seems correct to me. I'd like to understand if there was some logic for why we didn't do this check before to make sure we don't introduce a different error (left as a comment)

File: core/corehttp/gateway_handler.go (hunk 1 above)
Comment (aschmahmann) [reply core/corehttp/gateway_handler.go:399]: Also, how do we even reach the point where we error here? It seems like the UnixFS directory is telling us "here are the objects I have" and then we look for them and it errors.

Comment (Stebalien) [reply core/corehttp/gateway_handler.go:399]: > Also, how do we even reach the point where we error here? It seems like the UnixFS directory is telling us "here are the objects I have" and then we look for them and it errors.

Invalid sharded directory. Or a subshard of a sharded directory that we can list but can't traverse.

**Review by mvdan — APPROVED** [review]
Not knowing the codebase before, SGTM. The previous code was definitely open to panics.

I guess you could just do the shortHash change to avoid panics. I don't have enough context to tell if `err != nil` should continue or stop.

**Comment by Stebalien** [comment]
Ok, I'm going to merge this for now. We can change it later if there is a better fix.

**Pull request merged**